### PR TITLE
GSDX: Add TV shaders and some changes to dialog.

### PR DESCRIPTION
**Pull request by ssakash**
#### New Shader Configuration and main GSDX dialogs.
side main

**Summary of changes:-**

- Move custom shader settings to shade boost dialog and integrate them to form the Shader Configuration dialog. I decided to do something like this since the main GSDX dialog got way too crowded.

- Add a Combobox for TV Shaders on the Shader configuration dialog. Inspired by https://github.com/PCSX2/pcsx2/commit/a41308e8c318cde5617e7a1378b77fc13f605b85

Build download link: https://ci.appveyor.com/api/buildjobs/u0xy0rijhuhw3xa0/artifacts/pcsx2-v1.3.1-1476-g1f50811-GSDX_dlgchanges-251-vs2015-Win32-ReleaseAll.7z

Needs review from @turtleli

### CODE REVIEW

**Comment by turtleli** [comment]
What. On. Earth.

Spacing is inconsistent, groupboxes aren't aligned with each other, the groupbox clips the contrast slider. I'm no design expert, but clearly a lot more effort can be put into this.

@willkuer will love this.

**Comment by ssakash** [comment]
No worries, I'll update it. ;)  

Are those the only flaws on the dialogs right now ? Shouldn't really be a problem to fix them. I was more curious about users opinion on the integration of custom shader settings and shade boost to a single dialog. This dialog change was a bit hurried and does really need an update to patch things up.

About the latter comment, I have no problem leaving this to either you (or) @willkuer. ONLY if you feel you could do a better job :p 

Edit: Ahh, how the heck did I miss that group box conflicting with the contrast slider O.O , I'll make sure to do a neat update within two days. If you have any other stuffs for improvements, feel free to mention.

**Comment by willkuer** [comment]
I think this is really a good idea. Could reduce complexicity a bit.
The restore to default button probably only resets the shadeboost stuff or? So probably you either need to implement it for all controls or put it to the group box.

The width is the same as the main conf. window? I think one can make the shader windows narrower?
The ok button is centered in the main dialog but right aligned in the shader dialog. Do we need a cancel button?
Does the proposed grouping makes sense? I would propose to separate into internal and external shader. Internal would be shade boost and tv shaders - external the other ones.

FXAA is not a shader or? Probably it has to move somewhere else? Maybe next to 8-bit textures or into hw hacks?


(So i am the guy who likes aligned ui controls and you are the one for trailing whitespaces? Does this show our degree of autism?)

**Comment by karasuhebi** [comment]
I like it @ssakash, seems like a good place to put it. I think the window could be made narrower though, like @willkuer mentioned. On the topic of what @willkuer mentioned, I also think a Cancel button would be nice, and I agree that FXAA should be taken out of there; I don't think people will think to look for an AA option in the shader menu, even though technically FXAA is a shader-based PPAA method.

**Comment by turtleli** [comment]
FXAA is a shader. The shader stuff affects both hardware and software, so it doesn't really fit into either.

The entire shader dialog consists of custom shader settings, so the "Custom Shader Settings" groupbox label doesn't make sense. Perhaps the groupboxes could be removed (maybe).

The Reset defaults button - I think it should only apply to the shade boost settings. But then again, if someone wants default settings, they can just disable the shade boost itself. So the button could potentially be removed. Hmm.

Consider moving the software mode settings groupbox and contents down slightly now that the main dialog height isn't a concern.

Instead of a button opening the shader dialog I would prefer it as a tab (like Linux). That would be quite a bit more work though, so for now, let's just consider the current dialog idea and whether it's an improvement.

(And yes, I do have various OCDs. ;) )

**Comment by ssakash** [comment]
Cleaned it up a bit , any more stuffs to add ?

shot

**Comment by turtleli** [comment]
The spaces after the commas in GSdx.rc should probably be removed - I think the dialog editor will remove them the next time the file is overwritten.

My suggestions:
Move TV Shaders and Enable FXAA to the top.
Put the groupbox back in.
Centre the groupboxes - currently the shadeboost groupbox is slightly to the left.

**Comment by willkuer** [comment]
All elements are aligned on the left side but not on the right. One could adjust width to align them on the right as well. Maybe it will look different if the groupbox reappears.
The TVShader label text is not vertically aligned to the combobox text. It is the same for blending unit accuracy in the main configuration window. Normally one does not align containers but the inner text to increase readability (probably neglectable). This is done everywhere else if I see it correctly.

**Comment by karasuhebi** [comment]
So FXAA is definitely staying in the Shader window then? Putting it back in the main GSdx window is not being considered?

P.S. I don't care either way, I don't use FXAA. I just think it'd be easier to find in the main GSdx window since I don't think most users are aware that FXAA is shader-based AA.

**Comment by ssakash** [comment]
Take 2 on patching up the dialog.

take2


> The spaces after the commas in GSdx.rc should probably be removed - I think the dialog editor will remove them the next time the file is overwritten.

dialog editor ? what's that ? either way removed those spaces.

> (probably neglectable)

Yeah 

> I don't care either way, I don't use FXAA. I just think it'd be easier to find in the main GSdx window since I don't think most users are aware that FXAA is shader-based AA.

I don't think user awareness is a reasonable factor here, accuracy in option placements are more important.

**Comment by willkuer** [comment]
I think the question is if we should have a similar window for all AA stuff that is implemented in pcsx2 and where we put all stuff as FXAA, MSAA, SSAA independent on the exact implementation.

**Comment by turtleli** [comment]
Groupboxes still seem ever so slightly to the left. A screenshot of the main dialog would be nice as well. 

Dialog editor - It's accessible via the resource view window. Useful when making larger changes.

I think another dialog for AA stuff isn't necessary right now, but maybe that's just my opinion.

**Comment by karasuhebi** [comment]
>I don't think user awareness is a reasonable factor here, accuracy in option placements are more important.

Fair enough :)

**Comment by Giroom** [comment]
This looks great, will this be in the Linux version too?

**Comment by willkuer** [comment]
No. Windows only.

Linux has already the tv shaders implemented.

Unfortunately linux gui and windows gui diverged a bit.

**Comment by ssakash** [comment]
> Groupboxes still seem ever so slightly to the left. A screenshot of the main dialog would be nice as well. 

could you please give me the coordinates for the groupbox you have in your mind ? would be much faster this way :p 

here's the screenshot of main dialog:- 

main

**Comment by gregory38** [comment]
```
This looks great, will this be in the Linux version too?
```
What is the issue of my Linux GUI? 

You have more options than on windows and it is functional. However alignment of checkbox suck and so space between line.

**Comment by turtleli** [comment]
@ssakash
 - My mistake on the groupboxes being off centre - It looks off centre in your screenshot.
 - Rename the GSShadeBostDlg class to reflect the new purpose of the dialog. It's misspelt anyway. Also rename IDD_SHADEBOOST.
 - Remove the IDRESET define since it's no longer used..
 - Shade boost is a postprocessing shader - it does not apply only to hardware mode.
 - IDC_MIPMAP, IDC_PRELOAD_GS (from previous PR) and IDC_TVSHADER defines should actually be placed after IDC_UPSCALE_MULTIPLIER_TEXT.
 - Spaces are still present after commas in GSdx.rc (if you save with the dialog editor it should sort itself out).
 - Dialog editor margins are now messed up since you edited by hand.
 - Enable FXAA checkbox is overlong and extends outside of the dialog.
 - TV Shader label height is 14, every other label is 8.
 - The spacing inside the Shade Boost and External Shader groupboxes is inconsistent - compare the location of the "Enable Shade Boost" and "Enable External Shader" checkboxes.
 - The "Enable Shade Boost" and "Enable External Shader" checkboxes also do not have the same x coordinate.

Having a separate Shader Configuration Label and separate Configure button feels a bit redundant but I'm not sure what to do about that so I'll leave that alone (but someone else might have a better idea).

File: plugins/GSdx/GSSettingsDlg.cpp [hunk 1]
```diff
@@ -238,14 +222,7 @@ bool GSSettingsDlg::OnCommand(HWND hWnd, UINT id, UINT code)
...
 			break;
-		case IDC_SHADER_FX_BUTTON:
-			if (code == BN_CLICKED)
-				OpenFileDialog(IDC_SHADER_FX_EDIT, "Select External Shader");
-			break;
-		case IDC_SHADER_FX_CONF_BUTTON:
-			if (code == BN_CLICKED)
-				OpenFileDialog(IDC_SHADER_FX_CONF_EDIT, "Select External Shader Config");
-			break;
+
```
Comment (turtleli) [inline plugins/GSdx/GSSettingsDlg.cpp]: Remove blank line.

File: plugins/GSdx/GSSettingsDlg.cpp [hunk 2]
```diff
@@ -543,21 +523,60 @@ bool GSShadeBostDlg::OnMessage(UINT message, WPARAM wParam, LPARAM lParam)
...
 		{
-		case IDOK: 
+		case IDOK:
 		{
+			INT_PTR data;
+			//TV Shader
+			if (ComboBoxGetSelData(IDC_TVSHADER, data))
+				{
+					theApp.SetConfig("TVshader", (int)data);
+				}
```
Comment (turtleli) [inline plugins/GSdx/GSSettingsDlg.cpp]: Indentation issues.

**Comment by Giroom** [comment]
Nothing is wrong with it, It's just I completely missed that there are TV shaders somehow, I blame it on me using the windows version only until a couple months ago, so I assumed that every feature was present in both. My bad, I'll have to really look over all of the options some time.

**Comment by ssakash** [comment]
I did an update to the code, by the way @turtleli  if you don't mind, could you provide me the new required margin values for the dialogs ?  I'm having some stupid VS2013 specific conflicts with this dialog editor. I cleared the spaces manually. (if some spaces are still there, please mention)

> Enable FXAA checkbox is overlong and extends outside of the dialog.

I guess you meant the unnecessary higher width value ? if yes, should be sorted out, I tried aligning the FXAA checkbox a bit to the browse boxes but not really perfect due to limited width of the dialog and I'd really try to avoid increasing it further.

**Comment by turtleli** [comment]
Margins - I'll probably just update that myself.
Yes I did mean width.
Spaces - check mipmapping.
Shadeboost/External Shader Checkboxes are out of line with the other elements (x=14 for checkboxes, x=15 for everything else). Alternatively you could go for text alignment instead (which I think willkuer prefers).

Dialog looks like:
untitled

Anyone have any further comments on layout?

File: plugins/GSdx/GSSettingsDlg.cpp [hunk 3]
```diff
@@ -454,35 +407,41 @@ void GSSettingsDlg::UpdateControls()
...
-	EnableWindow(GetDlgItem(m_hWnd, IDC_SHADER_FX_BUTTON), external_shader_selected);
-	EnableWindow(GetDlgItem(m_hWnd, IDC_SHADER_FX_CONF_TEXT), external_shader_selected);
-	EnableWindow(GetDlgItem(m_hWnd, IDC_SHADER_FX_CONF_EDIT), external_shader_selected);
-	EnableWindow(GetDlgItem(m_hWnd, IDC_SHADER_FX_CONF_BUTTON), external_shader_selected);
-
-	// Shade Boost
-	EnableWindow(GetDlgItem(m_hWnd, IDC_SHADEBUTTON), IsDlgButtonChecked(m_hWnd, IDC_SHADEBOOST) == BST_CHECKED);
 }
 
 // Shade Boost Dialog
```
Comment (turtleli) [inline plugins/GSdx/GSSettingsDlg.cpp]: Comment needs updated/removed. Your call.

**Comment by ssakash** [comment]
> Margins - I'll probably just update that myself.
Spaces - check mipmapping.

Thanks 

> Shadeboost/External Shader Checkboxes are out of line with the other elements (x=14 for checkboxes, x=15 for everything else). Alternatively you could go for text alignment instead (which I think willkuer prefers).

actually the difference in X coordinate was intentional to show that the checkboxes control the enable/disable state of those other elements. either way, changed them to similar X coordinate. text alignment needs extra width for the dialog , so no :p

**Comment by ramapcsx2** [comment]
Just chiming in to say that this is nice work!
When the alignment stuff (so much nitpicking! :p) is done, it should get merged.

**Comment by ssakash** [comment]
Thanks Rama :) , here's how the current dialog looks:

gs1

Any stuffs which needs to be added guys ?

**Comment by turtleli** [comment]
It's probably fine. No one's complained that the ok and cancel buttons are slightly off centre. ;)

I'll likely merge tomorrow.

**Pull request merged**